File: Trees/bst.c
#include<stdio.h>
#include<stdlib.h>

struct Node{
    int data;
    struct Node *left;
    struct Node *right;
};

struct Node *createNode(int data){
    struct Node *node = (struct Node *)malloc(sizeof(struct Node));
    node->data = data;
    node->left = NULL;
    node->right = NULL;
}

struct Node *Insert(struct Node *root, int data){
  if(root == NULL){
    return createNode(data);
  }

  if(root->data > data){
    root->left = Insert(root->left, data);
  }
  else{
    root->right = Insert(root->right, data);
  }
  return root;
}

struct Node *Delete(struct Node *root, int val){
    if(root->data > val){
        root->left = Delete(root->left, val);
    }
    else if(root->data < val){
        root->right = Delete(root->right, val);
    }
    else{
        if(root->left == NULL && root->right == NULL){
            return NULL;
        }

        if(root->left == NULL){
            return root->right;
        }
        else if(root->right == NULL){
            return root->left;
        }

        struct Node *temp = inordersucessive(root->right);
        root->data = temp->data;
        root->right = Delete(root, temp->data);
    }
    return root;

}
struct Node *inordersucessive(struct Node *root){
    while(root->left != NULL){
        root = root->left;
    }
    return root;
}

File: Trees/bst.java
class Node{
    int val;
    Node right;
    Node left;
    Node(int val){
        this.val = val;
        right = null;
        left = null;
    }
}
public class bst {
    public static Node add(Node root, int data){
        if(root == null){
            Node node = new Node(data);
            return node;
        }
        else if(root.val>data){
            root.left = add(root.left, data);
        }
        else{
            root.right = add(root.right, data);

        }
        return root;
    }
    public static  void display(Node root){
          if(root == null){
            return;
          }
          System.out.print(root.val+" ");
          display(root.left);
          display(root.right);

    }
public static void search(Node root, int data){
    
    if(root == null){
        System.out.println("\nThere is no such value present");
      return;  
    }
    if(root.val == data){
        System.out.println("\nWe found an Element");
        return;
    }
    if(root.val>data){
        search(root.left, data);
    }
    else{
        search(root.right, data);
    }
}
public static Node delete(Node root, int data){
    if(root.val>data){
        root.left = delete(root.left, data);
    }
    else if(root.val<data){
        root.right = delete(root.right, data);
    }
    else{
        //case 1
        if(root.right == null && root.left == null){
            return null;
        }
        //case 2
        if(root.right == null){
            return root.right;
        }
        else if(root.left == null){
            return root.left;
        }
        //case 3
        Node temp = inordersucessive(root.right);
        root.val = temp.val;
        root.right = delete(root, temp.val);
    }
    return root;
}
public static Node inordersucessive(Node root){
    while(root.left != null){
        root = root.left;
    }
    return root;
}
    public static void main(String args[]){
         //creating root;
         Node root = new Node(4);
         add(root, 3);
         add(root, 2);
         add(root, 1);
         add(root, 5);
         add(root, 6);
         display(root);
         root = delete(root, 1);
        //  search(root, 7);
        System.out.println();
        display(root);

    }
}
